datashader/core.py

```diff
@@ -837,6 +835,14 @@ def quadmesh(self, source, x=None, y=None, agg=None):
             raise ValueError(f'DataArray name {source.name!r} does not match '
                              f'supplied reduction {agg}.')
 
+        if source[name].ndim == 3:
+            return self._quadmesh_3d(source, x, y, name, xarr, yarr, agg)
+        return self._quadmesh_2d(source, x, y, name, xarr, yarr, agg)
+
+    def _quadmesh_2d(self, source, x, y, name, xarr, yarr, agg):
+        """Process 2D quadmesh data."""
+        from .glyphs import QuadMeshRaster, QuadMeshRectilinear, QuadMeshCurvilinear
+
         if xarr.ndim == 1:
```

Comment:
this `xaxis, yaxis` stuff can be expensive for the curvilinear case; and should not repeated 3x. It seems possible to refactor so the coordinates are only inferred once?